sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/ClientRetryPolicy.java

```diff
@@ -173,8 +173,17 @@ private ShouldRetryResult shouldRetryOnSessionNotAvailable() {
             retryDelay = Duration.ofMillis(ClientRetryPolicy.RetryIntervalInMS);
         }
         this.retryContext = new RetryContext(this.failoverRetryCount, false);
-        return this.globalEndpointManager.refreshLocationAsync(null, forceRefresh)
-                .then(Mono.just(ShouldRetryResult.retryAfter(retryDelay)));
+        Mono<Void> completable = this.globalEndpointManager.refreshLocationAsync(null, forceRefresh);
+        if (isReadRequest || WebExceptionUtility.isWebExceptionRetriable(e)) {
+            // refresh cache and
+            // if it is a read request or if it is a write but we are sure the write
+            // hasn't reached the service retry
+            return completable.then(Mono.just(ShouldRetryResult.retryAfter(retryDelay)));
+        } else {
+            // refresh cache and
+            // no retry for writes which we are not sure if have reached to the service or not
+            return completable.then(Mono.just(ShouldRetryResult.noRetry()));
```

Comment:
Isn't heading of PR is contradicting , we are retrying on write in case of network failure ? Until I am missing anything

Reply:
We are only retrying if it is a readRequest, or WebExceptionRetriable. 
In the else case, we are not retrying.. we are calling noRetry()

Reply:
Exactly , but from heading “dont retry on network failure” Lets discuss offline

Reply:
Ok so it is retry on write in case of connection exception , thinking other way round was confusing.